fastcrypto-zkp/src/circom.rs

```diff
@@ -3,20 +3,106 @@
+
+/// A struct that can represent a Bn254 point serialized as a radix10 encoded string.
+#[derive(Debug, Clone, JsonSchema, Eq, PartialEq)]
+pub struct Bn254FieldElement([u8; 32]);
+
+impl Bn254FieldElement {
+    /// Returns the unpadded version of the point. This is the point with leading zeros removed.
+    pub fn unpadded(&self) -> &[u8] {
+        let mut buf = self.0.as_slice();
+
+        while !buf.is_empty() && buf[0] == 0 {
+            buf = &buf[1..];
+        }
+
+        // If the value is '0' then just return a slice of length 1 of the final byte
+        if buf.is_empty() {
+            &self.0[31..]
+        } else {
+            buf
+        }
+    }
+
+    /// Returns the padded version of the point. This is the point with leading zeros preserved to 32 bytes.
+    pub fn padded(&self) -> &[u8] {
+        &self.0
+    }
+}
+impl std::str::FromStr for Bn254FieldElement {
+    type Err = Bn254FieldElementParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let big_int = Fq::from_str(s).map_err(|_| Bn254FieldElementParseError::InvalidInput)?;
+        let be_bytes = BigUint::from_str(&big_int.to_string())?.to_bytes_be();
+        let len = be_bytes.len();
+        let mut buf = [0; 32];
+
+        if len > 32 {
+            return Err(Bn254FieldElementParseError::TooBig);
+        }
+
+        buf[32 - len..].copy_from_slice(&be_bytes);
+        Ok(Self(buf))
+    }
+}
+
+impl std::fmt::Display for Bn254FieldElement {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let big_int = num_bigint::BigUint::from_bytes_be(&self.0);
+        let radix10 = big_int.to_str_radix(10);
+        f.write_str(&radix10)
+    }
+}
+
+// Bn254Point's serialized format is as a radix10 encoded string
+impl Serialize for Bn254FieldElement {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.to_string().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Bn254FieldElement {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = std::borrow::Cow::<'de, str>::deserialize(deserializer)?;
+        std::str::FromStr::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Bn254FieldElementParseError {
+    #[error("unable to parse radix10 encoded value `{0}`")]
+    Parse(#[from] num_bigint::ParseBigIntError),
+    #[error("larger than 32 bytes")]
+    TooBig,
+    #[error("Invalid input")]
+    InvalidInput,
+}
 
 /// Parse a string as a field element in BN254. Return an `FastCryptoError::InvalidInput` error if
 /// the parsing fails.
-fn parse_field_element(s: &str) -> Result<Fq, FastCryptoError> {
-    s.parse::<Fq>().map_err(|_| FastCryptoError::InvalidInput)
+fn parse_field_element(s: &Bn254FieldElement) -> Result<Fq, FastCryptoError> {
+    s.to_string()
```

Comment:
Avoid the string conversion and do something like `Fq::from_be_bytes_mod_order(&s.0)` instead. This never fails now, so the return value doesn't have to be a `Result` anymore.